Add file to aid post processing

jtlau opened the pull request:
Hello,

I've been using this file (`combine_samples.C`) to load in multiple snapshots (in my case, from parametric simulations), subtract the mean, and compute the SVD over the snapshots.

This is similar but extended capability to `tests/load_samples.C`. 

I'm not sure the best location for this helper file / function, but it was simple to add it to `tests/`. 

I think it might be useful to anyone who needs to work with parametric simulations. Subtracting the mean is optional. There is more description and an example command in the file, included here:

Description: This is a file function that reads snapshot or basis files
		      and computes the SVD over the group. Also provides the option
		      to subtract the mean of the group before computing the SVD.

Example: `mpirun -n 2 ./combine_samples -b -m file1 file2 file3`

chldkdtn commented:
@dylan-copeland @kevinhkhuynh @siuwuncheung, it sounds like combine_samples.C does the similar thing to our offset capability. Let's discuss where is the best place to put this file and how to merge it to the existing architecture of libROM.

dylan-copeland commented:
Thank you @jtlau for this nice example of how to combine snapshot files. We have similar scripts in codes like Laghos, which we typically call "merge" scripts, that basically read snapshot files along with some offset vector and compute the SVD on the combined or merged snapshots. I don't think we have an example of how to do this in libROM, so adding this example seems helpful to other users. 

Regarding the "mean" vector, it appears to be a vector read from file, not necessarily the mean. The mean must be computed in a different code, so this example does not need to call it the mean. More generally, it can be called an "offset" vector.

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,186 @@
+    }
+    else {
+        // Close file so it can be opened:
+        static_basis_generator = nullptr;		
```
Instead of writing and reading the same data, can you just use `static_basis_generator->getSnapshotMatrix`?

jtlau replied:
Yes! I will change that.

jtlau commented:
Thanks @dylan-copeland for the comments. I am including an option to subtract an arbitrary offset file currently. Do you know what format the offset might likely be in? An hdf5 snapshot file?

dylan-copeland commented:
> Thanks @dylan-copeland for the comments. I am including an option to subtract an arbitrary offset file currently. Do you know what format the offset might likely be in? An hdf5 snapshot file?

Yes, hdf5 can be assumed. The libROM `Vector` class has the `read` and `write` functions for HDF files, which we use in Laghos for example.

jtlau commented:
@dylan-copeland I have included the option to subtract an offset, using the `Vector` `read` function. How does that look?

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
```
2022

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+ *
+ *****************************************************************************/
+
+// Description: This is a file function that reads snapshot or basis files
```
This function

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+    int dim   = 0;
+    bool subtract_mean = false;
+    bool subtract_offset = false;
+    std::string kind = "snapshot";
```
Not a big deal, but `kind` could be an `enum`.

dylan-copeland replied:
I see now that this string is used in `BasisGenerator`, so we can leave this as a string.

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+	    dim    = basis->numRows();
+	    snaps += basis->numColumns();
+	}
+    }
```
The verification of `dim > 0` could go here, along with setting another variable `dimFirst = dim` if `dimFirst` is not already set, otherwise verification that `dim == dimFirst`. This will ensure all files have the same dimensions, which could help users prevent or find mistakes.

jtlau replied:
Added both verifications

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+        static_basis_generator->loadSamples(sample_name, kind);
+    }
+
+    if (rank==0) std::cout << "Saving data uploaded as a snapshot" << std::endl;
```
as a snapshot matrix

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+    	int num_rows = snapshots->numRows();
+    	int num_cols = snapshots->numColumns();
+        CAROM::Vector *offset = new CAROM::Vector(num_rows, true);
+    	//double* offset = new double[num_rows];
```
remove

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+    if (rank==0) std::cout << "Saving data uploaded as a snapshot" << std::endl;
+    static_basis_generator->writeSnapshot();
+	
+    if (!subtract_mean && !subtract_offset) {
```
`CAROM_VERIFY(!(subtract_mean && subtract_offset));`

jtlau replied:
It could be possible that the user doesn't want to subtract the mean or offset, so I've included that possibility here. Unless did you mean that I should include that verification inside this if statement?

dylan-copeland replied:
I meant the user should not specify both mean and offset.

jtlau replied:
oh of course. Updated

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+            offset->read(offset_file);
+        }
+
+        /*-- Subtract mean from snapshot/bases matrix --*/
```
Replace mean with offset, for more generality? Or say "Subtract mean or offset"?

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+        }
+
+        /*-- Subtract mean from snapshot/bases matrix --*/
+    	CAROM::Matrix* snaps_mean = new CAROM::Matrix(num_rows, num_cols, false);
```
Instead of allocating another matrix, can you just subtract the offset from `snapshots`? For large matrices, this could be a significant amount of memory that is doubled.

jtlau replied:
Done

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+    	    for (int row = 0; row < num_rows; row++) {
+    	        snap_cur[row] = snaps_mean->item(row,col);
+    	    } 
+    	    static_basis_generator2->takeSample(snap_cur, 0.0, false);
```
It is unfortunate that another generator is necessary. Again, it seems 2x the necessary memory is being used. I don't see a good way to avoid this, unless we add the capability to modify existing snapshots. That is, to just subtract the offset from the snapshots already taken. Maybe this can be done in a future PR.

dylan-copeland reviewed (changes requested):
Thanks @jtlau for the option to subtract an offset. It looks good, and I just have some minor suggestions.

chldkdtn commented:
@dylan-copeland @kevinhkhuynh @siuwuncheung @jtlau, Can we think of combining "merge" in Laghos and "combineSnapshot" in SU2 into one script and include the script in libROM so that other physics solver can take advantage as well?

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,200 @@
+        CAROM::BasisReader reader(sample_name);
+		
+        if (kind == "snapshot") {
+	    CAROM::Matrix *snapshots = (CAROM::Matrix*) reader.getSnapshotMatrix(0);
```
It looks like this reads the entire matrix into memory, although you only need to get the dimensions here. The function `getSnapshotMatrix` first reads the dimensions and then reads the matrix. You could write a new function, `getSnapshotDimensions`, which just reads the dimensions. To avoid code duplication, `getSnapshotDimensions` could be called by `getSnapshotMatrix`. The same goes for `getSpatialBasis`.

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -91,22 +91,26 @@ int main(int argc, char* argv[])
     }
 	
     CAROM_VERIFY((snaps > 0) && (dim > 0));
+    CAROM_VERIFY(dim == dimFirst); // help to ensure files have the same dimensions
```
If this is inside the loop over `sample_names`, it will verify that all the files are of the same dimension.

jtlau replied:
I see, updated!

jtlau commented:
The remaining two comments (currently) from @dylan-copeland  require a separate PR implementing a few smaller capabilities within libROM (a dimension read and editing of snapshots). I will start a new PR for those changes - they should not be complicated.

jtlau commented:
Finally got around the finishing this up.

The last PR I submitted created the functions `getDim` and `getNumSamples`, but only for snapshot matrix files. I generalized the functions for spatial and temporal bases just now (see latest commit).

Dylan, for your remaining suggestion, I briefly looked into adding the capability to modify existing snapshots, but there didn't seem to be a simple solution, so I agree that would be better for a future and separate PR. 

Can you look over the recent changes? If it looks good I believe this PR is finished.

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,190 @@
+    	    generator_filename+"_sub"));
+    	
+    	for (int col = 0; col < num_cols; col++) {
+            CAROM::Vector *snap_cur = new CAROM::Vector(num_rows, true);
```
This allocates `snap_cur` for every column, without deleting it. It should be allocated before this loop over `col`. Also, it can be `CAROM::Vector snap_cur(num_rows, true);` so it doesn't have to be deleted.

dylan-copeland commented on tests/combine_samples.C:
```diff
@@ -0,0 +1,190 @@
+        
+    	int num_rows = snapshots->numRows();
+    	int num_cols = snapshots->numColumns();
+        CAROM::Vector *offset = new CAROM::Vector(num_rows, true);
```
This could be `CAROM::Vector offset(num_rows, true);` rather than using `new` and `delete`.

dylan-copeland commented:
@jtlau Thanks for the recent changes. I agree that we can postpone subtracting an offset from a basis generator's snapshots to another PR. For now, it would be helpful to add a brief comment about how that should be implemented in the future, to eliminate the need for the second generator.

Also, the `combine_samples.C` should be in the `examples` directory, not `tests` (which is for regression tests). Maybe you can make a new subdirectory `examples/misc` for this file. This file is a general example, without concrete input/output that should be tested. It is not necessary, but if you wanted to make a test, this file could go in `examples`, and most of the code in `main` could be put into a function in the library which could be called from a test in the `tests` directory. Also not necessary, you could add comments on how to generate some snapshot data from one of the other examples in libROM.

Finally, the code needs to be stylized, but the script `stylize.sh` is broken. To fix it, this version can be copied and added to this PR:
https://github.com/LLNL/libROM/blob/save_dmd_csv/scripts/stylize.sh
If you are not familiar with astyle and would like me to do this and run astyle, just let me know and I will commit the formatting changes.

dylan-copeland commented:
@jtlau Thanks for the recent changes. I agree that we can postpone subtracting an offset from a basis generator's snapshots to another PR. For now, it would be helpful to add a brief comment about how that should be implemented in the future, to eliminate the need for the second generator.

Also, the `combine_samples.C` should be in the `examples` directory, not `tests` (which is for regression tests). Maybe you can make a new subdirectory `examples/misc` for this file. This file is a general example, without concrete input/output that should be tested. It is not necessary, but if you wanted to make a test, this file could go in `examples`, and most of the code in `main` could be put into a function in the library which could be called from a test in the `tests` directory. Also not necessary, you could add comments on how to generate some snapshot data from one of the other examples in libROM.

Finally, the code needs to be stylized, but the script `stylize.sh` is broken. To fix it, this version can be copied and added to this PR:
https://github.com/LLNL/libROM/blob/save_dmd_csv/scripts/stylize.sh
If you are not familiar with astyle and would like me to do this and run astyle, just let me know and I will commit the formatting changes.

jtlau commented:
Thanks for the comments @dylan-copeland. I moved the function to `examples`, and added its own CMakeLists section so it will be compiled even when not using MFEM. I was able to stylize the file, too. 

I created a test in `tests`, although it's not quite what you described, but I think it is a simple way to test the function and provide an example. What do you think?

dylan-copeland reviewed (approved):
Thanks @jtlau for addressing all my comments. This PR looks good!
